base/abstractarraymath.jl

```diff
@@ -330,89 +308,147 @@ julia> repeat([1 2; 3 4], inner=(2, 1), outer=(1, 3))
+            Number of repetitions cannot be negative. Got inner = $inner
+            """
+            throw(ArgumentError(msg))
+        end
+        if length(inner) < ndims(arr)
+            msg = """
+            Number of dimensions of repetitions and array must match. Got
+            inner = $inner
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
+    end
+    if outer != nothing
+        if any(<(0), outer)
+            msg = """
+            Number of repetitions cannot be negative. Got outer = $outer
+            """
+            throw(ArgumentError(msg))
+        end
+        if (length(outer) < ndims(arr)) && (inner != nothing)
+            msg = """
+            Number of dimensions of repetitions and array must match. Got
+            outer = $outer
+            size(arr) = $(size(arr))
+            """
+            throw(ArgumentError(msg))
+        end
     end
+end
 
-    # fill the first inner block
-    if all(isequal(1), inner)
-        idxs = (axes(A)..., ntuple(n->OneTo(1), ndims(R)-ndims(A))...) # keep dimension consistent
-        R[idxs...] = A
-    else
-        inner_indices = [1:n for n in inner]
-        for c in CartesianIndices(axes(A))
-            for i in 1:ndims(A)
-                n = inner[i]
-                inner_indices[i] = (1:n) .+ ((c[i] - 1) * n)
-            end
-            fill!(view(R, inner_indices...), A[c])
+repeat_inner_outer(arr, inner::Nothing, outer::Nothing) = arr
+repeat_inner_outer(arr, ::Nothing, outer) = repeat_outer(arr, outer)
+repeat_inner_outer(arr, inner, ::Nothing) = repeat_inner(arr, inner)
+repeat_inner_outer(arr, inner, outer) = repeat_outer(repeat_inner(arr, inner), outer)
+
+function repeat_outer(a::AbstractMatrix, (m,n)::NTuple{2, Any})
+    o, p = size(a,1), size(a,2)
+    b = similar(a, o*m, p*n)
+    for j=1:n
+        d = (j-1)*p+1
+        R = d:d+p-1
+        for i=1:m
+            c = (i-1)*o+1
+            @inbounds b[c:c+o-1, R] = a
         end
     end
+    return b
+end
 
-    # fill the outer blocks along each dimension
-    if all(isequal(1), outer)
-        return R
+function repeat_outer(a::AbstractVector, (m,)::Tuple{Any})
+    o = length(a)
+    b = similar(a, o*m)
+    for i=1:m
+        c = (i-1)*o+1
+        @inbounds b[c:c+o-1] = a
     end
-    src_indices  = [1:n for n in inner_shape]
-    dest_indices = copy(src_indices)
-    for i in eachindex(outer)
-        B = view(R, src_indices...)
-        for j in 2:outer[i]
-            dest_indices[i] = dest_indices[i] .+ inner_shape[i]
-            R[dest_indices...] = B
+    return b
+end
+
+function repeat_outer(arr::AbstractArray{<:Any,N}, dims::NTuple{N,Any}) where {N}
+    insize  = size(arr)
+    outsize = map(*, insize, dims)
+    out = similar(arr, outsize)
+    for I in CartesianIndices(arr)
+        for J in CartesianIndices(dims)
+            TIJ = map(Tuple(I), Tuple(J), insize) do i, j, d
+                i + d * (j-1)
```

Comment:
Ahh, but outer repeats did previously work (but are unfortunately untested):

```Julia
# master
julia> repeat(OffsetArray(rand(3,3),-1,-1), outer=(2,1))
6×3 Matrix{Float64}:
 0.625501  0.113374  0.643544
 0.870337  0.994211  0.464454
 0.518634  0.999477  0.040367
 0.625501  0.113374  0.643544
 0.870337  0.994211  0.464454
 0.518634  0.999477  0.040367

# this PR, currently:
julia> repeat_outer(OffsetArray(rand(3,3),-1,-1), (2,1))
Segmentation fault: 11
```

Reply:
A segfault is clearly bad and I need to fix it. The question is, what is the most reasonable semantics for an OffsetArray? I think returning an OffsetArray does not make sense, since it is unclear what should happen with the axes. That leaves throwing an error or silently passing to one based indexing. I am in favor of the former, but that might be considered a breaking change.

Reply:
Funny, something must have been wrong with my REPL state when I tested that (I was just doing simplistic copy/paste debugging at the REPL for a quick review).  Now that I try this again, it looks like your implementation matches the status quo — outer repeats work (returning a 1-based array) and inner repeats go out of bounds.

We could simply `Base.require_one_based_indexing` for the inner repeat case and preserve the outer's status quo, leaving it for another day... but I agree the returned axes are questionable.